coral/functions/licence_number_function.py

```diff
@@ -0,0 +1,275 @@
+
+            if not status_tile:
+                app_id = (
+                    tile.data.get(SYSTEM_REF_RESOURCE_ID_NODE).get("en").get("value")
+                )
+                # Set the licence name to use the app id
+                Tile.objects.get_or_create(
+                    resourceinstance_id=resource_instance_id,
+                    nodegroup_id=LICENCE_NAME_NODEGROUP,
+                    data={
+                        LICENCE_NAME_NODE: {
+                            "en": {
+                                "direction": "ltr",
+                                "value": f"Excavation Licence {app_id}",
+                            }
+                        }
+                    },
+                )
+                return
+
+        if tile_nodegroup_id == STATUS_NODEGROUP:
+            if tile.data.get(STATUS_NODE) != STATUS_FINAL_VALUE:
+                return
+            try:
+                cur_d_tile = Tile.objects.get(
+                    resourceinstance_id=resource_instance_id,
+                    nodegroup_id=CUR_D_DECISION_NODEGROUP,
+                )
+                if cur_d_tile.data.get(CUR_D_DECISION_NODE) != CUR_D_DECISION_APPROVED_VALUE:
+                    return
+            except Tile.DoesNotExist:
+                return
+            
+        if tile_nodegroup_id == CUR_D_DECISION_NODEGROUP:
+            if tile.data.get(CUR_D_DECISION_NODE) != CUR_D_DECISION_APPROVED_VALUE:
+                return
+            try:
+                status_tile = Tile.objects.get(
+                    resourceinstance_id=resource_instance_id,
+                    nodegroup_id=STATUS_NODEGROUP,
+                )
+                if status_tile.data.get(STATUS_NODE) != STATUS_FINAL_VALUE:
+                    return
+            except Tile.DoesNotExist:
+                return
+
+        licence_number = generate_licence_number(resource_instance_id)
+
+        if not licence_number:
+            return
+
+        try:
+            # Configure the licence number
+            Tile.objects.get_or_create(
+                resourceinstance_id=resource_instance_id,
+                nodegroup_id=LICENCE_NUMBER_NODEGROUP,
+                data={
+                    LICENCE_NUMBER_NODE: {
+                        "en": {"direction": "ltr", "value": licence_number}
+                    },
+                },
+            )
+            # Configure the licence name with the licence number included
+            licence_name_tile = Tile.objects.get(
+                resourceinstance_id=resource_instance_id,
+                nodegroup_id=LICENCE_NAME_NODEGROUP,
+            )
+            licence_name_tile.data[LICENCE_NAME_NODE]["en"][
+                "value"
+            ] = f"Excavation Licence {licence_number}"
+            licence_name_tile.save()
+        except Exception as e:
+            print(f"Failed saving licence number external ref or licence name: {e}")
+            raise e
+        
+        try:
```

Comment:
I added this try catch where it adds the licence as an associate licence for the activity